src/SSHDebugPS/Docker/DockerConnection.cs

```diff
@@ -0,0 +1,110 @@
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using Microsoft.VisualStudio.Debugger.Interop.UnixPortSupplier;
+
+namespace Microsoft.SSHDebugPS.Docker
+{
+    internal class DockerConnection : PipeConnection
+    {
+        private string _containerName;
+
+        public DockerConnection(DockerTransportSettings settings, Connection outerConnection, string name, string containerName)
+            : this(settings, outerConnection, name, containerName, null)
+        { }
+
+        public DockerConnection(DockerTransportSettings settings, Connection outerConnection, string name, string containerName, int? timeout)
+        : base(settings, outerConnection, name, timeout)
+        {
+            _containerName = containerName;
+        }
+
+        public override void BeginExecuteAsyncCommand(string commandText, bool runInShell, IDebugUnixShellCommandCallback callback, out IDebugUnixShellAsyncCommand asyncCommand)
+        {
+            if (IsClosed)
+            {
+                throw new ObjectDisposedException(nameof(PipeConnection));
+            }
+
+            if (runInShell)
+            {
+                AD7UnixAsyncShellCommand command = new AD7UnixAsyncShellCommand(CreateShellFromSettings(TransportSettings, OuterConnection), callback, closeShellOnComplete: true);
+                command.Start(commandText);
+
+                asyncCommand = command;
+            }
+            else
+            {
+                /// TODO: figure out what RunInShell means?
+                DockerExecSettings settings = new DockerExecSettings(_containerName, commandText, true);
+                AD7UnixAsyncCommand command = new AD7UnixAsyncCommand(CreateShellFromSettings(settings, OuterConnection, true), callback, closeShellOnComplete: true);
+
+                asyncCommand = command;
+            }
+        }
+
+        /// <summary>
+        /// Docker uses its own command system to copy files and directories into and out of the container.
+        /// </summary>
+        /// <param name="sourcePath"></param>
+        /// <param name="destinationPath"></param>
+        public override void CopyDirectory(string sourcePath, string destinationPath)
+        {
+            Copy(sourcePath, destinationPath);
+        }
+
+        public override void CopyFile(string sourcePath, string destinationPath)
+        {
+            Copy(sourcePath, destinationPath);
+        }
+
+        private void Copy(string source, string destination)
+        {
+            DockerCopySettings settings;
+            string tmpFile = null;
+
+            if(!Directory.Exists(source) && !File.Exists(source))
+            {
+                throw new ArgumentException(FormattableString.Invariant($"Local path: '{source}' does not exist"), nameof(source));
+            }
+
+            if (OuterConnection != null)
+            {
+                tmpFile = "/tmp" + "/" + Path.GetRandomFileName();
```

Comment:
>/tmp [](start = 27, length = 4)

I feel like '/tmp' is not the right place for this since it is a shared namespace, and we aren't doing anything to protect it. Maybe use `$XDG_RUNTIME_DIR` and fallback to ~/.cache if that doesn't exist.

Reply:
`$XDG_CACHE_HOME` resolves to ~/.cache by default, maybe use that and fall back to /tmp for non-interactive users that don't have a home directory?

---
In reply to: [253714264](https://github.com/Microsoft/MIEngine/pull/831#discussion_r253714264) [](ancestors = 253714264)

Reply:
I'm planning on using `Microsoft.VisualStudio.DockerPS.FileCopy.<GUID>` and leaving it in the `/tmp` directory. 

---
In reply to: [254075353](https://github.com/Microsoft/MIEngine/pull/831#discussion_r254075353) [](ancestors = 254075353,253714264)